pgmpy/base/MAG.py

```diff
@@ -0,0 +1,295 @@
+from typing import Hashable, Iterable, Optional
+
+import networkx as nx
+
+from pgmpy.base import AncestralBase
+
+
+class MAG(AncestralBase):
+    """
+    Class for representing Maximal Ancestral Graphs (MAGs).
+
+    A MAG is a type of graph used in causal inference to represent conditional
+    independence relations when some variables are latent (unobserved). Unlike
+    simple directed acyclic graphs (DAGs), MAGs allow for special edge types
+    (directed and bidirected) that capture the presence of latent confounding
+    and selection bias. Every pair of nodes in a MAG is connected in such a way
+    that the graph is "maximal," meaning no additional edges can be added
+    without changing the set of implied conditional independence relations.
+
+    References
+    ----------
+        [1] Zhang, J. (2008). Causal Reasoning with Ancestral Graphs. Journal of Machine Learning Research, 9(7).
+    """
+
+    def __init__(
+        self,
+        ebunch: Optional[Iterable[tuple[Hashable, Hashable]]] = None,
+        latents: set[Hashable] = set(),
+        roles=None,
+    ):
+        """
+        Initialize a Maximal Ancestral Graph
+
+        Parameters
+        ----------
+        ebunch : iterable of tuples, optional
+            A list or iterable of edges to add at initialization.
+
+        latents : set, default=set()
+            Set of latent (unobserved) variables.
+
+        roles : dict, optional (default: None)
+            A dictionary mapping roles to node names.
+            The keys are roles, and the values are role names (strings or iterables of str).
+            If provided, this will automatically assign roles to the nodes in the graph.
+            Passing a key-value pair via ``roles`` is equivalent to calling
+            ``with_role(role, variables)`` for each key-value pair in the dictionary.
+
+        Returns
+        -------
+        MAG
+            A new instance of a Maximal Ancestral Graph.
+
+        Examples
+        --------
+        >>> from pgmpy.base import MAG
+        >>> mag = MAG(
+        ...     ebunch=[("L", "A", "-", ">"), ("B", "C", "-", ">")], latents={"L"}
+        ... )
+        >>> sorted(mag.nodes())
+        ['A', 'B', 'C', 'L']
+
+        Roles can be assigned to nodes in the graph at construction or using methods.
+
+        At construction:
+
+        >>> mag = MAG(
+        ...     ebunch=[("L", "A", "-", ">"), ("B", "C", "-", ">")],
+        ...     latents={"L"},
+        ...     roles={"exposure": "A", "outcome": "B"},
+        ... )
+
+        Roles can also be assigned after creation using ``with_role`` method.
+
+        >>> mag = mag.with_role("adjustment", {"L", "C"})
+
+        Vertices of a specific role can be retrieved using ``get_role`` method.
+
+        >>> mag.get_role("exposure")
+        ["A"]
+        >>> mag.get_role("adjustment")
+        ["L", "C"]
+
+        """
+        super().__init__(ebunch=ebunch, latents=latents, roles=roles)
```

Comment:
We need to check the edge types here before calling super. Not all edge types are allowed in MAGs.